File: homework/hw9/queen.cpp
#include <iostream>
#include <vector>

using namespace std;

const int N = 8; // 棋盤大小
vector < vector<int> > board(N, vector<int>(N, 0)); // 棋盤

// 檢查皇后是否可以放在(row, col)位置
bool isSafe(int row, int col) {
    // 檢查同一列
    for (int i = 0; i < row; i++) {
        if (board[i][col] == 1) {
            return false;
        }
    }

    // 檢查左上對角線
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    // 檢查右上對角線
    for (int i = row, j = col; i >= 0 && j < N; i--, j++) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    return true;
}

// 解決八皇后問題的遞歸函數
bool solveNQueens(int row) {
    if (row == N) {
        // 所有皇后都被成功放置，打印棋盤
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                cout << board[i][j] << " ";
            }
            cout << endl;
        }
        return true;
    }

    for (int col = 0; col < N; col++) {
        if (isSafe(row, col)) {
            board[row][col] = 1;
            if (solveNQueens(row + 1)) {
                return true;
            }
            board[row][col] = 0; // 回溯
        }
    }
    return false;
}

int main() {
    if (solveNQueens(0)) {
        cout << "" << endl;
    } else {
        cout << "無法找到解。" << endl;
    }

    return 0;
}

File: homework/hw9/queen.py
Size = 8
count=0
board = [[0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0],
        [0,0,0,0,0,0,0,0]]

def B_print(board):
    for row in board:
        print(row)

def Is_Safe(row,col):
    for r in range(row):
        if board[r][col]==1:return 0
    
    c=col
    for r in range(row-1,-1,-1):
        c-=1
        if c<0: break
        if board[r][c]==1:return 0

    c=col
    for r in range(row-1,-1,-1):
        c+=1
        if c>=Size: break
        if board[r][c]==1:return 0
    return 1

def dfs(row):
    global count
    #print('board:')
    #B_print(board)
    if row==Size: 
        count+=1
        print(f'Solution {count} :')
        B_print(board)
        return 1
    for c in range(Size):
        if (Is_Safe(row,c)):
            board[row][c]=1
            #B_print(board)
            dfs(row+1)
            board[row][c]=0


dfs(0)